### ISSUE TITLE
Dependency conflicts on io.netty:netty-common, leading to inconsistent program behaviors

### ISSUE THREAD

**Issue by HelloCoCooo** [salesforce/reactive-grpc#226]
Hi, in **reactive-grpc/common/reactive-grpc-benchmarks**, there are mulptiple versions of library **io.netty:netty-common**. However, according to Maven's dependency management strategy: **_"first declaration wins"_**, only **io.netty:netty-common:4.1.33.Final** can be loaded, and **io.netty:netty-common:4.1.38.Final** will be shadowed.

**In total, there are 4 conflicting API pairs between these two library version.**

As shown in the following figure, your project expects to invoke method **_<io.netty.util.AbstractReferenceCounted: refCnt()I>_**  in library _**io.netty:netty-common:4.1.38.Final**_ (along the _original dependency path_). As it has been shadowed, this method defined in **_io.netty:netty-common:4.1.33.Final_** is actually forced to be referenced via the following invocation path (along the _actual dependency path_):

```
<com.salesforce.reactivegrpc.jmh.proto.Control$ScenarioResult: getSerializedSize()I> /home/wwww/sensor/unzip/reactive-grpc-1.0.1/common/reactive-grpc-benchmarks/target/classes
<io.netty.handler.codec.HeadersUtils$1: get(I)Ljava/lang/Object;> /home/wwww/.m2/repository/io/netty/netty-codec/4.1.38.Final/netty-codec-4.1.38.Final.jar
<io.netty.handler.codec.HeadersUtils$1: get(I)Ljava/lang/String;> /home/wwww/.m2/repository/io/netty/netty-codec/4.1.38.Final/netty-codec-4.1.38.Final.jar
<io.netty.util.ReferenceCountUtil$ReleasingTask: toString()Ljava/lang/String;> /home/wwww/.m2/repository/io/netty/netty-common/4.1.33.Final/netty-common-4.1.33.Final.jar
<io.netty.handler.ssl.ReferenceCountedOpenSslContext: refCnt()I> /home/wwww/.m2/repository/io/netty/netty-handler/4.1.38.Final/netty-handler-4.1.38.Final.jar
<io.netty.util.AbstractReferenceCounted: refCnt()I>
```

![reactive-grpc](https://user-images.githubusercontent.com/46306510/93980449-38128d00-fdb1-11ea-8a4c-a3d7431f8310.png)

Although both of these conflicting libraries contain the referenced methods (**_with the same signature_**), they have **_different implementations_**. This issue will not cause runtime crashes, but it can introduce inconsistent semantic program hehaviors----

Code snippet of **_<io.netty.util.AbstractReferenceCounted: refCnt()I>_** in _**io.netty:netty-common:4.1.38.Final**_  (shadowed but expected to invoke method):

<details><summary>detailed method body</summary>

```java
public int refCnt() {
        return updater.refCnt(this);
    }


//instance : updater
public final int refCnt(T instance) {
        return realRefCnt(updater().get(instance));
    }

private static int realRefCnt(int rawCnt) {
        return rawCnt != 2 && rawCnt != 4 && (rawCnt & 1) != 0 ? 0 : rawCnt >>> 1;
    }
```

</details>

Code snippet of **_<io.netty.util.AbstractReferenceCounted: refCnt()I>_** in _**io.netty:netty-common:4.1.33.Final**_ (loaded version):

<details><summary>detailed method body</summary>

```java
public int refCnt() {
        return realRefCnt(refCntUpdater.get(this));
    }

private static int realRefCnt(int rawCnt) {
        return (rawCnt & 1) != 0 ? 0 : rawCnt >>> 1;
    }
```

</details>

**The detailed informantion of the remaining 3 conflicting API pairs can be found in the following attachment.**
[4 conflicting API pairs in project reactive-grpc-benchmarks.txt](https://github.com/salesforce/reactive-grpc/files/5428306/4.conflicting.API.pairs.in.project.reactive-grpc-benchmarks.txt)

## **Dependency tree--**

[INFO] com.salesforce.servicelibs:reactive-grpc-benchmarks:jar:1.0.1
[INFO] +- io.reactivex.rxjava2:rxjava:jar:2.2.12:compile
[INFO] |  \- org.reactivestreams:reactive-streams:jar:1.0.2:compile
[INFO] +- com.salesforce.servicelibs:reactor-grpc-stub:jar:1.0.1:compile
[INFO] |  +- com.salesforce.servicelibs:reactive-grpc-common:jar:1.0.1:compile
[INFO] |  |  +- javax.annotation:javax.annotation-api:jar:1.3.2:compile
[INFO] |  |  +- (org.reactivestreams:reactive-streams:jar:1.0.2:compile - omitted for duplicate)
[INFO] |  |  \- (io.grpc:grpc-stub:jar:1.23.0:compile - omitted for duplicate)
[INFO] |  \- io.projectreactor:reactor-core:jar:3.2.12.RELEASE:compile
[INFO] |     \- (org.reactivestreams:reactive-streams:jar:1.0.2:compile - omitted for duplicate)
[INFO] +- com.salesforce.servicelibs:rxgrpc-stub:jar:1.0.1:compile
[INFO] |  +- (com.salesforce.servicelibs:reactive-grpc-common:jar:1.0.1:compile - omitted for duplicate)
[INFO] |  \- (io.reactivex.rxjava2:rxjava:jar:2.2.12:compile - omitted for duplicate)
[INFO] +- io.grpc:grpc-netty:jar:1.23.0:compile
[INFO] |  +- (io.grpc:grpc-core:jar:1.23.0:compile - omitted for duplicate)
[INFO] |  +- io.netty:netty-codec-http2:jar:4.1.38.Final:compile
[INFO] |  |  +- **(io.netty:netty-common:jar:4.1.38.Final:compile - omitted for conflict with 4.1.33.Final)**
[INFO] |  |  +- (io.netty:netty-buffer:jar:4.1.38.Final:compile - omitted for conflict with 4.1.33.Final)
[INFO] |  |  +- io.netty:netty-transport:jar:4.1.38.Final:compile
[INFO] |  |  |  +- **(io.netty:netty-common:jar:4.1.38.Final:compile - omitted for duplicate)**
[INFO] |  |  |  +- (io.netty:netty-buffer:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |  |  \- io.netty:netty-resolver:jar:4.1.38.Final:compile
[INFO] |  |  |     \- (io.netty:netty-common:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |  +- io.netty:netty-codec:jar:4.1.38.Final:compile
[INFO] |  |  |  +- **(io.netty:netty-common:jar:4.1.38.Final:compile - omitted for duplicate)**
[INFO] |  |  |  +- (io.netty:netty-buffer:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |  |  \- (io.netty:netty-transport:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |  +- io.netty:netty-handler:jar:4.1.38.Final:compile
[INFO] |  |  |  +- (io.netty:netty-common:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |  |  +- (io.netty:netty-buffer:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |  |  +- (io.netty:netty-transport:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |  |  \- (io.netty:netty-codec:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |  \- io.netty:netty-codec-http:jar:4.1.38.Final:compile
[INFO] |  |     +- **(io.netty:netty-common:jar:4.1.38.Final:compile - omitted for duplicate)**
[INFO] |  |     +- (io.netty:netty-buffer:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |     +- (io.netty:netty-transport:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |     +- (io.netty:netty-codec:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  |     \- (io.netty:netty-handler:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |  \- io.netty:netty-handler-proxy:jar:4.1.38.Final:compile
[INFO] |     +- **(io.netty:netty-common:jar:4.1.38.Final:compile - omitted for duplicate)**
[INFO] |     +- (io.netty:netty-buffer:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |     +- (io.netty:netty-transport:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |     +- (io.netty:netty-codec:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |     +- io.netty:netty-codec-socks:jar:4.1.38.Final:compile
[INFO] |     |  +- **(io.netty:netty-common:jar:4.1.38.Final:compile - omitted for duplicate)**
[INFO] |     |  +- (io.netty:netty-buffer:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |     |  +- (io.netty:netty-transport:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |     |  \- (io.netty:netty-codec:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] |     \- (io.netty:netty-codec-http:jar:4.1.38.Final:compile - omitted for duplicate)
[INFO] +- io.grpc:grpc-core:jar:1.23.0:compile
[INFO] |  +- io.grpc:grpc-api:jar:1.23.0:compile
[INFO] |  |  +- io.grpc:grpc-context:jar:1.23.0:compile
[INFO] |  |  +- com.google.errorprone:error_prone_annotations:jar:2.3.2:compile
[INFO] |  |  +- com.google.code.findbugs:jsr305:jar:3.0.2:compile
[INFO] |  |  +- org.codehaus.mojo:animal-sniffer-annotations:jar:1.17:compile
[INFO] |  |  \- (com.google.guava:guava:jar:26.0-android:compile - omitted for duplicate)
[INFO] |  +- com.google.code.gson:gson:jar:2.7:compile
[INFO] |  +- com.google.android:annotations:jar:4.1.1.4:compile
[INFO] |  +- io.perfmark:perfmark-api:jar:0.17.0:compile
[INFO] |  |  +- (com.google.code.findbugs:jsr305:jar:3.0.2:compile - omitted for duplicate)
[INFO] |  |  \- (com.google.errorprone:error_prone_annotations:jar:2.3.2:compile - omitted for duplicate)
[INFO] |  +- io.opencensus:opencensus-api:jar:0.21.0:compile
[INFO] |  \- io.opencensus:opencensus-contrib-grpc-metrics:jar:0.21.0:compile
[INFO] |     \- (io.opencensus:opencensus-api:jar:0.21.0:compile - omitted for duplicate)
[INFO] +- io.grpc:grpc-stub:jar:1.23.0:compile
[INFO] |  \- (io.grpc:grpc-api:jar:1.23.0:compile - omitted for duplicate)
[INFO] +- io.grpc:grpc-protobuf:jar:1.23.0:compile
[INFO] |  +- (io.grpc:grpc-api:jar:1.23.0:compile - omitted for duplicate)
[INFO] |  +- com.google.protobuf:protobuf-java:jar:3.9.0:compile
[INFO] |  +- com.google.guava:guava:jar:26.0-android:compile
[INFO] |  |  +- org.checkerframework:checker-compat-qual:jar:2.5.2:compile
[INFO] |  |  \- com.google.j2objc:j2objc-annotations:jar:1.1:compile
[INFO] |  +- com.google.api.grpc:proto-google-common-protos:jar:1.12.0:compile
[INFO] |  \- io.grpc:grpc-protobuf-lite:jar:1.23.0:compile
[INFO] |     +- (io.grpc:grpc-api:jar:1.23.0:compile - omitted for duplicate)
[INFO] |     \- (com.google.guava:guava:jar:26.0-android:compile - omitted for duplicate)
[INFO] +- io.netty:netty-buffer:jar:4.1.33.Final:compile
[INFO] |  \- **io.netty:netty-common:jar:4.1.33.Final:compile**
[INFO] +- org.openjdk.jmh:jmh-core:jar:1.21:compile
[INFO] |  +- net.sf.jopt-simple:jopt-simple:jar:4.6:compile
[INFO] |  \- org.apache.commons:commons-math3:jar:3.2:compile
......

## Suggested solutions:

**Solution1:** Update direct dependency io.netty:netty-buffer from 4.1.33.Final to 4.1.38.Final. Because it(io.netty:netty-buffer：4.1.38.Final) uses the dependency io.netty:netty-common:4.1.38.Final.

**Solution2:**  Declare version io.netty:netty-common:4.1.38.Final as a direct dependency, to override the version 4.1.33.Final (based on Maven's nearest wins loading strategy).

**Solution3:** Remove the direct dependency io.netty:netty-buffer:4.1.33.Final. Use transitive dependencies from io.grpc:grpc-netty:1.23.0.

Thanks.
Best regards,
Coco

**Comment by HelloCoCooo** [comment]
@rmichela Could please help me check this issue?
May I pull a request to fix it?
Thanks again.

**Comment by rmichela** [comment]
Solution 3 is best. We always want the version of Netty gRPC is using. 

Thanks so much for submitting this. Please submit a PR.

Was this issue found with a code scanning tool?

**Comment by HelloCoCooo** [comment]
I implemented an automated tool, which combines static analysis and dynamic testing to detect such potencial semantic behavior changes due to dependency conflicts.
By static analysis we can identify the dependency conflict issues and compare the differences of control flows and data flows between the shadowed method and loaded method (with the same signature).
Then, for the isomerous method pairs, we use test cases to see the impacts of such differences between shadowed method and loaded method on client project's semantic behaviors.

Are you interested in the detection method of dependency conflicts?

**Comment by rmichela** [comment]
I'd love to see your scanner. Is there a github repo I can check out?

**Comment by rmichela** [comment]
Released in 1.1.0
